ARCs/arc-0060.md

```diff
@@ -0,0 +1,233 @@
+
+#### `Errors`
+
+These are the possible errors that the wallet **MUST** handle:
+
+| Error | Description |
+| --- | --- |
+| `ERROR_INVALID_SCOPE` | The `scope` is not valid. |
+| `ERROR_FAILED_DECODING` | The `data` field could not be decoded. |
+| `ERROR_INVALID_SIGNER` | Unable to find in the wallet the public key related to the signer. |
+| `ERROR_MISSING_DOMAIN` | The `domain` field is missing. |
+| `ERROR_MISSING_DOMAIN` | The `authenticatedData` field is missing. |
+| `ERROR_BAD_JSON` | The `data` field is not a valid JSON object. |
+| `ERROR_FAILED_DOMAIN_AUTH` | The `authenticatedData` field does not match the hash of the `domain`. |
+| `ERROR_FAILED_HD_PATH` | The `hdPath` field is not a valid BIP44 path. |
+
+## Backwards Compatibility
+
+N / A
+
+## Reference Implementation
+
+Available in the `assets/arc-0060` folder.
+
+### Sample Use cases
+
+#### Generic AUTH
+
+```ts
+    const authData: Uint8Array = new Uint8Array(createHash('sha256').update("arc60.io").digest())
+
+    const authRequest: StdSigData = {
+        data: Buffer.from("{[jsonfields....]}").toString('base64'),
+        signer: publicKey,
+        domain: "arc60.io",
+        requestId: Buffer.from(randomBytes(32)).toString('base64'),
+        authenticationData: authData,
+        hdPath: "m/44'/60'/0'/0/0"
+    }
+
+    const signResponse = await arc60wallet.signData(authRequest, { scope: ScopeType.AUTH, encoding: 'base64' })
+```
+
+#### CAIP-122
+
+```ts
+    const caip122Request: CAIP122 = {
+        domain: "arc60.io",
+        chain_id: "283",
+        account_address: ...
+        type: "ed25519",
+        statement: "We are requesting you to sign this message to authenticate to arc60.io",
+        uri: "https://arc60.io",
+        version: "1",
+        nonce: Buffer.from(randomBytes(32)).toString,
+        ...
+    }
+
+    // Disply message title according EIP-4361
+    const msgTitle: string = `Sign this message to authenticate to ${caip122Request.domain} with account ${caip122Request.account_address}`
+
+    // Display message body according EIP-4361
+    const msgBodyPlaceHolders: string = `URI: ${caip122Request.uri}\n` + `Chain ID: ${caip122Request.chain_id}\n`
+        + `Type: ${caip122Request.type}\n`
+        + `Nonce: ${caip122Request.nonce}\n`
+        + `Statement: ${caip122Request.statement}\n`
+        + `Expiration Time: ${caip122Request["expiration-time"]}\n`
+        + `Not Before: ${caip122Request["not-before"]}\n`
+        + `Issued At: ${caip122Request["issued-at"]}\n`
+        + `Resources: ${(caip122Request.resources ?? []).join(' , \n')}\n`
+
+    // Display message according EIP-4361
+    const msg: string = `${msgTitle}\n\n${msgBodyPlaceHolders}`
+    console.log(msg)
+
+    // authenticationData
+    const authenticationData: Uint8Array = new Uint8Array(createHash('sha256').update(caip122Request.domain).digest())
+
+    const signData: StdSigData = {
+        data: Buffer.from(JSON.stringify(caip122Request)).toString('base64'),
```

Comment:
Is `JSON.stringify` deterministic? Since JSON objects are unordered, we might have to resort to something like [canonical JSON](https://www.rfc-editor.org/rfc/rfc8785) to derive a deterministic byte-array from JSON data suitable for signing.

Reply:
Yup; great point. Should be deterministic JSON

Reply:
Updated @k13n